### DISCUSSION TITLE
Version 2.3.1

### DISCUSSION THREAD

**Discussion by saran-t** [google-deepmind/mujoco#625] [Announcements]
## Python bindings

1. The `simulate` GUI is now available through the `mujoco` Python package as `mujoco.viewer`. See [documentation](https://mujoco.readthedocs.io/en/2.3.1/python.html#pyviewer) for details. (Contribution by [Levi Burner](https://github.com/aftersomemath).)
1. The ``Renderer`` class from the MuJoCo tutorial Colab is now available directly in the native Python bindings.

## General

3. The tendon `springlength` attribute can now take two values. Given two non-decreasing values, `springlength` specifies a [deadband](https://en.wikipedia.org/wiki/Deadband) range for spring stiffness. If the tendon length is between the two values, the force is 0. If length is outside this range, the force behaves like a regular spring, with the spring resting length corresponding to the nearest `springlength` value. This can be used to create tendons whose limits are enforced by springs rather than constraints, which are cheaper and easier to analyse. See [tendon_springlength.xml](https://github.com/deepmind/mujoco/tree/main/test/engine/testdata/tendon_springlength.xml) example model.  _This is a minor breaking API change. `mjModel.tendon_lengthspring` now has size `ntendon x 2` rather than  `ntendon x 1`._

1. Removed the requirement that stateless actuators come before stateful actuators.
1. Added [mju_fill](https://mujoco.readthedocs.io/en/2.3.1/APIreference.html#mju-fill), [mju_symmetrize](https://mujoco.readthedocs.io/en/2.3.1/APIreference.html#mju-symmetrize) and [mju_eye](https://mujoco.readthedocs.io/en/2.3.1/APIreference.html#mju-eye) utility functions.
1. Added `gravcomp` attribute to [body](https://mujoco.readthedocs.io/en/2.3.1/XMLreference.html#body) implementing gravity compensation and buoyancy. See [balloons.xml](https://github.com/deepmind/mujoco/tree/main/model/balloons/balloons.xml) example model.
1. Renamed the `cable` plugin library to `elasticity`.
1. Added `actdim` attribute to [general actuators](https://mujoco.readthedocs.io/en/2.3.1/XMLreference.html#actuator-general). Values greater than 1 are only allowed for `user`, as native activation dynamics are all scalar. Added example test implementing 2nd-order activation dynamics to [engine_forward_test.cc](https://github.com/deepmind/mujoco/blob/main/test/engine/engine_forward_test.cc).
1. Improved particle [composite](https://mujoco.readthedocs.io/en/2.3.1/XMLreference.html#body-composite) type, which now permits a user-specified geometry and multiple joints. See the two new examples: [particle_free.xml](https://github.com/deepmind/mujoco/tree/main/model/composite/particle_free.xml) and [particle_free2d.xml](https://github.com/deepmind/mujoco/tree/main/model/composite/particle_free2d.xml).
1. Performance improvements for non-AVX configurations:

   - 14% faster `mj_solveLD` using [restrict](https://en.wikipedia.org/wiki/Restrict>). See [engine_core_smooth_benchmark_test](https://github.com/deepmind/mujoco/tree/main/test/benchmark/engine_core_smooth_benchmark_test.cc).
    - 50% faster `mju_dotSparse` using manual loop unroll. See [engine_util_sparse_benchmark_test](https://github.com/deepmind/mujoco/tree/main/test/benchmark/engine_util_sparse_benchmark_test.cc).
1. Added new `solid` passive force plugin.
    - This is new force field compatible with the [composite](https://mujoco.readthedocs.io/en/2.3.1/XMLreference.html#body-composite) particles.
    - Generates a tetrahedral mesh having particles with mass concentrated at vertices.
    - Uses a piecewise-constant strain model equivalent to finite elements but expressed in a coordinate-free
      formulation. This implies that all quantities can be precomputed except edge elongation, as in a mass-spring model.
    - Only suitable for small strains (large displacements but small deformations). Tetrahedra may invert if subject to
      large loads.

1. Added API functions [mj_loadPluginLibrary](https://mujoco.readthedocs.io/en/2.3.1/APIreference.html#mj-loadPluginLibrary) and  [mj_loadAllPluginLibraries](https://mujoco.readthedocs.io/en/2.3.1/APIreference.html#mj-loadAllPluginLibraries). The first function is identical to `dlopen` on a POSIX system, and to `LoadLibraryA` on Windows. The second function scans a specified directory for all dynamic libraries file and loads each library found. Dynamic libraries opened by these functions are assumed to register one or more MuJoCo plugins on load.
1. Added an optional `visualize` callback to plugins, which is called during `mjv_updateScene`. This callback allows custom plugin visualizations. Enable stress visualization for the Cable plugin as an example.
1. Sensors of type [user](https://mujoco.readthedocs.io/en/2.3.1/XMLreference.html#sensor-user) no longer require `objtype`, `objname` and `needstage`. If unspecified, the objtype is now [mjOBJ_UNKNOWN](https://mujoco.readthedocs.io/en/2.3.1/APIreference.html#mjtobj). `user` sensors `datatype` default is now `real`, `needstage` default is now `acc`.
1. Added support for capsules in URDF import.
1. On macOS, issue an informative error message when run under [Rosetta 2](https://support.apple.com/en-gb/HT211861) translation on an Apple Silicon machine. Pre-built MuJoCo binaries make use of [AVX](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions) instructions on x86-64 machines, which is not supported by Rosetta 2. (Before this version, users only get a cryptic "Illegal instruction" message.)

## Bug fixes

17. Fixed bug in `mj_addFileVFS` that was causing the file path to be ignored (introduced in 2.1.4).

## Simulate

18. Renamed the directory in which the `simulate` application searches for plugins from `plugin` to `mujoco_plugin`.
1.  Mouse force perturbations are now applied at the selection point rather than the body center of mass.

<hr /><em>This discussion was created from the release <a href='https://github.com/deepmind/mujoco/releases/tag/2.3.1'>2.3.1</a>.</em>

**Comment by mhming** [comment]
when I build the simulate folder, I got some error
![image](https://user-images.githubusercontent.com/21165493/209084753-4f0a61e5-80b0-4f47-85d9-a1cc1d9b4c37.png)
after I add the lodepng.h, the error is:
![image](https://user-images.githubusercontent.com/21165493/209084898-896d0417-8ccd-430d-9770-44f62c4a00a2.png)

the right way is:
1. add the lodepng.cpp & lodepng.h in simulaye folder
2.  change the makefile as follow:

COMMON=-O2 -I../include -L../lib -std=c++17 -pthread -Wl,-no-as-needed -Wl,-rpath,'$$ORIGIN'/../lib

all:
	$(CXX) $(COMMON)  -c simulate.cc uitools.cc glfw_dispatch.cc lodepng.cpp
	$(CXX) $(COMMON) main.cc simulate.o uitools.o glfw_dispatch.o lodepng.o -lmujoco -lglfw -o ../bin/simulate
	rm uitools.o simulate.o glfw_dispatch.o lodepng.o
